File: sdks/python/trace_harness/loading/facts.py
"""Business-owned dependency declarations; the harness owns their execution."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass

from trace_harness.model.context import TraceContext
from trace_harness.model.node import Node


@dataclass(frozen=True)
class EvidenceDependency:
    span_ids: tuple[str, ...]
    fields: tuple[str, ...] | None


@dataclass(frozen=True)
class FactDependency:
    node_id: str
    name: str


@dataclass(frozen=True)
class FactProducer:
    produces: tuple[str, ...]
    applies: Callable[[Node], bool]
    requires: Callable[[Node, TraceContext], tuple[EvidenceDependency | FactDependency, ...]]
    compute: Callable[[Node, TraceContext], dict]
    version: str = "1"

File: sdks/typescript/trace-harness/src/loading/facts.ts
import type { TraceContext } from "../model/context";
import type { Node } from "../model/node";

export class EvidenceDependency {
  constructor(readonly span_ids: readonly string[], readonly fields: readonly string[] | null) {}
}
export class FactDependency {
  constructor(readonly node_id: string, readonly name: string) {}
}
export type Dependency = EvidenceDependency | FactDependency;
/** Business declares evidence and pure computation; the runtime schedules reads. */
export interface FactProducer {
  produces: readonly string[];
  applies(node: Node): boolean;
  requires(node: Node, trace: TraceContext): readonly Dependency[];
  compute(node: Node, trace: TraceContext): Record<string, unknown>;
}
